pyogrio/geopandas.py

```diff
@@ -226,13 +252,44 @@ def write_dataframe(
     # TODO: may need to fill in pd.NA, etc
     field_data = [df[f].values for f in fields]
 
-    if geometry_type is None:
-        geometry_type = "Unknown"
+    # Determine layer_geometry_type and/or promote_to_multi
+    if layer_geometry_type is None or promote_to_multi is None:
+        tmp_layer_geometry_type = "Unknown"
+
+        # If there is data, infer layer geometry type + promote_to_multi
         if not df.empty:
-            # TODO: validate geometry types, not all combinations are valid
             geometry_types = geometry.type.unique()
             if len(geometry_types) == 1:
-                geometry_type = geometry_types[0]
+                tmp_layer_geometry_type = geometry_types[0]
+                if promote_to_multi and tmp_layer_geometry_type in (
+                    "Point",
+                    "LineString",
+                    "Polygon",
+                ):
+                    tmp_layer_geometry_type = f"Multi{tmp_layer_geometry_type}"
+            elif len(geometry_types == 2):
+                # Check if the types are corresponding multi + single types
+                if "Polygon" in geometry_types and "MultiPolygon" in geometry_types:
+                    multi_type = "MultiPolygon"
+                elif (
+                    "LineString" in geometry_types
+                    and "MultiLineString" in geometry_types
+                ):
+                    multi_type = "MultiLineString"
+                elif "Point" in geometry_types and "MultiPoint" in geometry_types:
+                    multi_type = "MultiPoint"
+                else:
+                    multi_type = None
+
+                # If they are corresponding multi + single types
+                if multi_type is not None and driver in DRIVERS_NO_MIXED_SINGLE_MULTI:
+                    if promote_to_multi is None:
```

Comment:
Although it might not matter much in practice, as both geojson and shapefile (two example drivers that we test that could cover this) don't really care about the geometry type

Reply:
It indeed doesn't matter for the current supported formats, but it should indeed be like that. good catch!